File: 290.WordPattern/pattern.cc
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        map<char, int> indexP;
        map<string, int> indexS;
        int i = 0, n = pattern.size();
        stringstream in(str);
        for (string word; in >> word; i++) {
            if (indexP[pattern[i]] != indexS[word]) {
                return false;
            }
            indexP[pattern[i]] = indexS[word] = i+1;
        }
        return i == n;
    }
};

File: 290.WordPattern/pattern.py
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        strs = str.split(' ')
        if (len(pattern) != len(strs)):
            return False

        index = {}
        for i in range(len(pattern)):
            lastP = index.get(pattern[i]+'#', -1)
            lastS = index.get(strs[i], -1)
            if lastP != lastS:
                return False
            index[pattern[i]+'#'] = i
            index[strs[i]] = i

        return True

if __name__ == '__main__':
    pattern = 'abc'
    str = 'b c a'
    print Solution().wordPattern(pattern, str)
